components/brave_wallet/browser/eth_tx_controller_unittest.cc

```diff
@@ -5,13 +5,310 @@
+                                             const std::string& error_message) {
+  EXPECT_TRUE(success);
+  EXPECT_FALSE(id.empty());
+  EXPECT_TRUE(error_message.empty());
+  *callback_called = true;
+  *tx_meta_id = id;
+}
+
+void AddUnapprovedTransactionFailureCallback(bool* callback_called,
+                                             bool success,
+                                             const std::string& id,
+                                             const std::string& error_message) {
+  EXPECT_FALSE(success);
+  EXPECT_TRUE(id.empty());
+  EXPECT_FALSE(error_message.empty());
+  *callback_called = true;
+}
+
+class EthTxControllerUnitTest : public testing::Test {
+ public:
+  EthTxControllerUnitTest()
+      : task_environment_(base::test::TaskEnvironment::TimeSource::MOCK_TIME),
+        browser_context_(new content::TestBrowserContext()),
+        shared_url_loader_factory_(
+            base::MakeRefCounted<network::WeakWrapperSharedURLLoaderFactory>(
+                &url_loader_factory_)) {}
+
+  void SetUp() override {
+    url_loader_factory_.SetInterceptor(base::BindLambdaForTesting(
+        [&](const network::ResourceRequest& request) {
+          url_loader_factory_.ClearResponses();
+
+          base::StringPiece request_string(request.request_body->elements()
+                                               ->at(0)
+                                               .As<network::DataElementBytes>()
+                                               .AsStringPiece());
+          absl::optional<base::Value> request_value =
+              base::JSONReader::Read(request_string);
+          std::string* method = request_value->FindStringKey("method");
+          ASSERT_TRUE(method);
+
+          if (*method == "eth_estimateGas") {
+            url_loader_factory_.AddResponse(
+                request.url.spec(),
+                "{\"jsonrpc\":\"2.0\",\"id\":1,\"result\":"
+                "\"0x00000000000009604\"}");
+          } else if (*method == "eth_gasPrice") {
+            url_loader_factory_.AddResponse(
+                request.url.spec(),
+                "{\"jsonrpc\":\"2.0\",\"id\":1,\"result\":"
+                "\"0x595698248593c0000\"}");
+          }
+        }));
+
+    user_prefs::UserPrefs::Set(browser_context_.get(), &prefs_);
+    brave_wallet::RegisterProfilePrefs(prefs_.registry());
+    rpc_controller_.reset(
+        new EthJsonRpcController(shared_url_loader_factory_, &prefs_));
+    keyring_controller_.reset(new KeyringController(&prefs_));
+
+    auto tx_state_manager = std::make_unique<EthTxStateManager>(
+        &prefs_, rpc_controller_->MakeRemote());
+    auto nonce_tracker = std::make_unique<EthNonceTracker>(
+        tx_state_manager.get(), rpc_controller_.get());
+    auto pending_tx_tracker = std::make_unique<EthPendingTxTracker>(
+        tx_state_manager.get(), rpc_controller_.get(), nonce_tracker.get());
+
+    eth_tx_controller_.reset(new EthTxController(
```

Comment:
You can use factories to get all of these and no need to create the intermediate things like `nonce_tracker` manually in the test.  See `browser/brave_wallet/brave_wallet_provider_impl_unittest.cc` for an example.

Reply:
It seems I would need to create my own rpc_controller and hence all these classes to use the desired url_loader_factory to intercept calls?

Reply:
ok you can leave as is in that case